### ISSUE TITLE
SpO2 sensor not connecting to digital port

### ISSUE THREAD

**Issue by JamilAhmad1000** [biosignalsplux/cpp-samples#3]
Hello!
I am trying to connect two sensors, breathing sensor (analog port) and SpO2 sensor (digital port). However, I am unable to receive any data from the digital port, in fact, the SpO2 sensor works fine in opensignals software but when i try to record data with this code, it doesn't work. Any kind of help is appreciated.

**Comment by DFNOsorio** [comment]
Hello,

To use the SPO2 sensor you have to start your acquisition using a Source instead of the port number. Below is an example of the configuration for the SpO2 sensor and the other sensor you want to record data from. 

Plux::Source SpO2Source;
sourceIMU.port = 9;
sourceIMU.chMask = 0x03
sourceIMU.nBits = 16;
sourceIMU.freqDivisor = 1;

Plux::Source RIPSource;
sourceIMU.port = YourAnalogPort;
sourceIMU.chMask = 0x01
sourceIMU.nBits = 16;
sourceIMU.freqDivisor = 1;


Plux::Sources sources;
sources.push_back(SpO2Source);
sources.push_back(RIPSource);

dev.start(1000, sources);
dev.loop();
dev.stop();

**Comment by JamilAhmad1000** [comment]
Thank you so much for the response. 
I have a few questions if you can answer.
What do you mean by using the source? 
My respiratory sensor is working fine in all the analog ports, so i will just skip the RIP code from what you shared.
Can you please tell me where should i add this specific code that you have shared?

PS: if you can edit the files and upload the new versions, it will help the community a lot

**Comment by DFNOsorio** [comment]
Hi, 

I recommend the following reading to better understand the different ways to start an acquisition using the API:

https://www.downloads.plux.info/apis/cpp-api-docs/struct_plux_1_1_source.html (Source)
https://www.downloads.plux.info/apis/cpp-api-docs/class_plux_1_1_signals_dev.html#a9463dcfd11510aa57f3c659968e54ffc (Start with Sources)
https://github.com/pluxbiosignals/python-samples/blob/master/OneDeviceSpecialChannelsExample.py#L71-L84 (Python Example)

**Comment by JamilAhmad1000** [comment]
Hello,
I have tried implementing the code. The code works apart from the fact that I am unable to start acquisition from source. it prompts as error which says "simpleApp.cpp:38:25: error: no matching function for call to ‘SimpleDev::start(int, Plux::Sources&)’
   38 |   dev.start(1000,sources);". Any kind of help or guidance is appreciated

**Comment by DFNOsorio** [comment]
SimpleDev is an abstraction to the SignalsDev class of the API. To use sources, you have to change the SimpleDev files to include the start method using sources.

**Comment by JamilAhmad1000** [comment]
Hi,
The line for starting from source is present in plux.h (void start (float
basefreq, const Source &sources). i have included it in simpleDev.cpp and
simpleDev.h. however, it is still not working.

**Comment by JamilAhmad1000** [comment]
Hello!
I have made some progress regarding the SpO2 sensor. I have integrated the SpO2 code. I did not run into any errors. However, I am still not getting any output on digital port. In fact, the SpO2 sensor is off and isn't even turned on.

**Comment by GRamosPlux** [comment]
Dear **@JamilAhmad1000**,

Thank you very much for your valuable feedback, the team is really glad for understanding that you are making progresses.

Regarding the issue that persists, in order to deeply evaluate its potential source, would it be possible to share with the team:

- The current version of the **C++** script (complete or only the relevant segment where you interact with the **PLUX** device) that you are implementing;
- An **Acquisition File** generated by **OpenSignals (r)evolution** where it is used the same channel/device setup being tested in the **C++** script.

On the other hand, the team would like to check if the **Respiration** data continues to be properly received with the most recent changes.

We will carefully wait for your important reply, wishing, in the meantime, a continuation of a positive day!

**My best and sincere regards,**
@GRamosPlux

**Comment by JamilAhmad1000** [comment]
Hello!
First of all thank you so much for helping me out.
Certainly I can provide you with the code and the changes that I have made.

**Comment by JamilAhmad1000** [comment]
#include "simpleDev.h"

#include <string> //// If you want to define the mac address in a variable

int main() {
	try
	{
		///// Use this block to search for PLUXs devices (via BTH or USB)
		SimpleDev::VDevInfo devs = SimpleDev::find();
		for (int i = 0; i < devs.size(); i++) {
			printf("%s - %s\n", devs[i].path.c_str(), devs[i].description.c_str());
		}

		///// Use this block to connect to a device

		const std::string MAC_Addres = "BTH00:07:80:0F:2F:CC";   /// Devices mac address
		//const std::string MAC_Addres = "USB0";										/// Use the UBS adapter (Only for windows)

		SimpleDev dev(MAC_Addres);
		puts("Connected to device. Press Enter to exit.");
		std::string ver = dev.version();    // get device version string
		printf("Device version: %s\n", ver.c_str());
		
		/*****************************************************/
		
		Plux::Source SpO2Source;
		{
			SpO2Source.port = 9;
			SpO2Source.freqDivisor = 1;
			SpO2Source.nBits = 16;
			SpO2Source.chMask = 0x03;
		}
		
		Plux::Sources sources;	
		{
			sources.push_back(SpO2Source);
		}
		
		dev.start(1000, sources);
		
		dev.start(1000, { 1, 2, 3, 4, 5, 6, 7, 8 }, 16);	/// Start the device with sampling rate = 1000, all the channels and at 16bit resolution
		
		dev.trigger(true);
		
		/****************************************************/
		
		SimpleDev::VFrame frames(100); // initialize the frames vector with 100 frames (From simpleDev cpp)
		int nSamples = 0;
		int maxSamples = 4000;
		
		do
		{
			dev.read(frames); // get 100 frames from device
			const SimpleDev::Frame &f = frames[0];  // get a reference to the first frame of each 100 frames block
			printf("%d : %d ; %d %d %d %d %d %d %d %d\n",   // dump the first frame
				f.seq,
				f.digital[0],
				f.analog[0], f.analog[1], f.analog[2], f.analog[3],
				f.analog[4], f.analog[5], f.analog[6], f.analog[7]);
			nSamples += 100;
			} while (nSamples <= maxSamples); // until a key is pressed
		dev.stop(); // stop acquisition
	} // dev is destroyed here (it goes out of scope)
	catch (Plux::Exception &e)
	{
	printf("Exception: %s\n", e.getDescription());
	}
	return 0;
}

**Comment by JamilAhmad1000** [comment]
So i am using 2 sensors. Respiratory and SpO2 sensor. What I am trying to do here is that I don't want to specify an specific port for my Respiratory sensor. So i can just plug my sensor in any port from (1-8) to give me more freedom and it works. Secondly, I am also trying to start the SpO2 sensor using the digital port (port 9). Unfortunately, the SpO2 sensor doesn't even start (which can be clear if you see a red light on sensor). Also, I don't run into any errors when I run this code when I compile it and it makes it difficult for me to point out the problem.

**Comment by JamilAhmad1000** [comment]
I have added this code in simpleDev.cpp to start from source.
void SimpleDev::start(float baseFreq, const Sources &sources) 
{
		sigDev.start(baseFreq, sources);
}

And I have added this code in simpleDev.h file.
void start(float baseFreq, const Sources &sources);

I am not very good at c++. so if you see any mistake, let me know

**Comment by GRamosPlux** [comment]
Dear @JamilAhmad1000,

The team is grateful for your valuable feedback and for your availability to share the custom script.

Everything seems perfect, there are only two small details that we can explore:

- Comment the line of code `dev.start(1000, { 1, 2, 3, 4, 5, 6, 7, 8 }, 16);`;
- Update/Edit the print instruction to the following structure: `printf("%d : %d ; %d %d %d %d %d %d %d %d | SpO2 %d %d\n", f.seq, f.digital[0], f.analog[0], f.analog[1], f.analog[2], f.analog[3], f.analog[4], f.analog[5], f.analog[6], f.analog[7], f.analog[8], f.analog[9]);`

The team will carefully wait for your important message 😄 

**My best and sincere regards,**
@GRamosPlux

**Comment by JamilAhmad1000** [comment]
I am sorry but I don't think you have understood the problem. Even if I comment the line that you have mentioned, I am just starting the digital port (9). How am I suppose to start the analog ports with dev.start(1000, sources)?

**Comment by GRamosPlux** [comment]
Dear **@JamilAhmad1000**,

This is a very pertinent question, thank you very much for alerting me regarding the **Analog** ports detail.

Essentially, in order to initialize the **RIP** sensor port/source, the following version of the code should be considered:

```cpp
Plux::Source SpO2Source;
SpO2Source.port = 9;
SpO2Source.freqDivisor = 1;
SpO2Source.nBits = 16;
SpO2Source.chMask = 0x03;

Plux::Source RIPSource;
sourceIMU.port = 1; // If the port where the RIP sensor is connected is the number 1.
sourceIMU.chMask = 0x01
sourceIMU.nBits = 16;
sourceIMU.freqDivisor = 1;
	
Plux::Sources sources;	
sources.push_back(RIPSource);
sources.push_back(SpO2Source);
```

Then, if only these two ports are being used (**1** and **9**), the print instruction should present the following format:

```cpp
printf("%d : %d ; RIP %d SpO2 (RED) %d SpO2 (INFRARED) %d\n", f.seq, f.digital[0], f.analog[0], f.analog[1], f.analog[2]);
```

**My best and sincere regards,**
@GRamosPlux

**Comment by JamilAhmad1000** [comment]
this is exactly what I want to know. let's say I am using multiple sensors instead of just using two sensors. And I have to use all the analog ports (1-8) and also digital port (9). In this case, It won't be optimized code because I have to repeat the same code (port, chmask, nbits, freq) for all the sensors that I want to use. Instead, what I want to do is keep this freedom with me to turn on all the ports with the same frequency and then it is my choice if I use one sensor for analog port or multiple sensors. Also, I guess there is no need to start analog ports from the source as is shown in the base code.

**Comment by GRamosPlux** [comment]
Dear **@JamilAhmad1000**,

Thank you very much for the valuable feedback.

Indeed, as you wisely suggested, the repetition of the same lines of code (`chMask`, `freqDivisor`, `nBits`) is not the most elegant, but, fortunately, you don't really need to include these lines of codes, considering that the **default** values (**0x01**, **1**, **16**) will be automatically assigned when not specified by the developer.

Alternatively, if you consider viable, it is also possible to create a function that receives the `port`, `chMask`, `freqDivisor` and `nBits` as inputs and returns the respective `Source` (in order to prevent repeating the code).

Regarding the pertinent topic _"I guess there is no need to start analog ports from the source as is shown in the base code."_, unfortunately, the **API** would ignore the second `start` command, i.e., if a **Digital** sensor is being used, the **Sources** path will need to be followed in all active ports (**Analog** and **Digital**), guaranteeing that we only call a single `start` command.

I am pleased to share the full documentation of the **API** (the root link from which the ones shared by my colleague **@DFNOsorio** derive), which will give you more freedom and the possibility to edit/expand the **SimpleDev** class to meet your requirements 😊 

- https://www.downloads.plux.info/apis/cpp-api-docs/namespace_plux.html

**My best and sincere regards,**
@GRamosPlux

**Comment by JamilAhmad1000** [comment]
thank you so much for the response.
Unfortunately, after doing all the updates that you shared with me, the digital port is still not working (somehow), in fact the the analog port is also not working anymore with this new changes. it prompts this:
BTH00:07:80:0F:2F:CC - biosignalsplux
Connected to device. Press Enter to exit.
Device version: biosignalsplux v3.8
Exception: P�YFV

**Comment by GRamosPlux** [comment]
Dear **@JamilAhmad1000**,

The team is really sorry for understanding that the reported obstacle still persists.

A few moments ago, with the valuable help of my colleague **@DFNOsorio** we were able to test the **Sources** path using the code shared in one of our previous interactions and reshared below:

> ```c++
> Plux::Source SpO2Source;
> SpO2Source.port = 9;
> SpO2Source.freqDivisor = 1;
> SpO2Source.nBits = 16;
> SpO2Source.chMask = 0x03;
> 
> Plux::Source RIPSource;
> sourceIMU.port = 1; // If the port where the RIP sensor is connected is the number 1.
> sourceIMU.chMask = 0x01
> sourceIMU.nBits = 16;
> sourceIMU.freqDivisor = 1;
> 	
> Plux::Sources sources;	
> sources.push_back(RIPSource);
> sources.push_back(SpO2Source);
> ```
> 
> Then, if only these two ports are being used (**1** and **9**), the print instruction should present the following format:
> 
> ```c++
> printf("%d : %d ; RIP %d SpO2 (RED) %d SpO2 (INFRARED) %d\n", f.seq, f.digital[0], f.analog[0], f.analog[1], f.analog[2]);
> ```

The team was able to collect data from the **RIP** and **SpO2** sensors simultaneously.
It is possible that the error results from a specificity of your code, considering that the returned exception seems to not be directly related with the **API**.

Sadly, we are unable to replicate the problem and provide more detailed and reliable information.
Currently, our only suggestion will be to **1)** download again the **API** and its respective **Sample APP** available in the current repository; **2)** implement the **Sources** version of the `start` method; **3)** Provide the sources exactly as defined in the above quote; **4)** call the **Sources** version of the `start` method and **5)** rebuild/run the APP.

With this new iteration, we can restrict the changes applied to the code and eventually, avoid the change that was triggering the reported error.

**My best and sincere regards,**
@GRamosPlux

**Comment by JamilAhmad1000** [comment]
I am really sorry for bothering you again and again and again. 
I am doing exactly as you asked me to do. I think the problem is with my simpleDev.cpp and simpleDev.h file. I have edited them for the source. I have added this line of code in simpleDev.cpp file which is as follow:
-------------------------------------------------------------------------------
void SimpleDev::start(float baseFreq, const Sources &sources)
{
		sigDev.start(baseFreq, sources);
}
---------------------------------------------------------------------------
and I have added this line of code in my simpleDev.h file which is as follow:
-------------------------------------------------------------------------
typedef std::vector<Plux::Source> Sources;

void start(float baseFreq, const Sources &sources); (I believe something is wrong with the typedef, I am not sure)
-----------------------------------------------------------------------
And the funny thing is, I dont get any errors :(

**Comment by JamilAhmad1000** [comment]
As instructed by you, I have downloaded the repository again, made the changes as shown by you and the error still persists. it is not an error but the exception still persists. But before the exception the SpO2 sensor blinks only once and then it prompts the exception. I have shared all the changes in the above comment that I have made. I haven't changed anything else other than adding the above lines of code. 
However, I have managed to print the exception by changing the exception part in the code a bit. The exception says:
Exception: Invalid operation on current device state (code 0).

**Comment by JamilAhmad1000** [comment]
Hi,
The exception is still there even after the changes you suggested. I started from scratch and it is still there.
Connected to device. Press Enter to exit.
Device version: biosignalsplux v3.8
Exception: Invalid operation on current device state (code 0).

**Comment by GRamosPlux** [comment]
Dear **@JamilAhmad1000**,

Firstly, I would like to sincerely apologize for my late reply.

Regarding the new tests that you carefully conducted, thank you very much for your attentive update, it is always a pleasure to help 😊 

My colleague **@DFNOsorio** alerted me regarding a detail that may be relevant in the new function that you wisely created.
The team recommends the inclusion of the following line of code:

```cpp
void SimpleDev::start(float baseFreq, const Sources &sources)
{
     sigDev.start(baseFreq, sources);
     // Add the following line of code (considering that you are only using the RIP and SpO2 sources).
     sigDev.nChannels = 3;
}
```

I will carefully wait for your always important feedback!

**My best and sincere regards,**
@GRamosPlux

**Comment by JamilAhmad1000** [comment]
Hello and Good Afternoon!
I hope you are doing well.
The code is finally working and I am able to receive data from Analog Port and Digital Port.
Thank you so much for the assistance and help provided by you and your team (I thank me for believing in me :D)
Thank you so much 👍

**Comment by GRamosPlux** [comment]
Dear **@JamilAhmad1000**,

Fortunately, everything is well with the team, thank you very much for your kind wishes.
We sincerely wish that your day is also being pleasant.

Regarding the **API** challenge, we are really glad for understanding that you were able to overtake the reported obstacle 🎉 
Your perseverance was definitively the secret that allowed you to achieve the glory in this journey 😊 

I will now close the current issue 😉 

Please, if any new question arises, do not hesitate to open a [**New Issue**](https://github.com/pluxbiosignals/cpp-samples/issues/new).
Meanwhile, it is with great pleasure that the entire team wishes you a continuation of an excellent day!

**My best and sincere regards,**
@GRamosPlux
